File: Tests/SpeechXTests/TypeOverDetectorTests.swift
import XCTest
@testable import SpeechX

final class TypeOverDetectorTests: XCTestCase {

    private func corrected(_ injected: [String], _ baseline: String, _ current: String) -> String? {
        TypeOverDetector.correction(
            injectedWords: Set(injected.map { $0.lowercased() }),
            baseline: baseline, current: current
        )?.corrected
    }

    // MARK: - Should detect a correction

    func testBasicSpellingFix() {
        XCTAssertEqual(corrected(["tell", "jon", "about", "it"], "tell jon about it", "tell john about it"), "john")
    }

    func testPreservesCorrectedWordCasing() {
        XCTAssertEqual(corrected(["jon"], "hi Jon", "hi John"), "John")
    }

    // Regression: count-based diff must survive repeated words in the field.
    func testCorrectedWordAlreadyPresentElsewhere() {
        XCTAssertEqual(corrected(["ayewsh"], "hi ayush ayewsh", "hi ayush ayush"), "ayush")
    }

    func testInjectedWordRepeatsEditingOneOccurrence() {
        XCTAssertEqual(corrected(["ayewsh"], "ayewsh ayewsh", "ayush ayewsh"), "ayush")
    }

    // MARK: - Should NOT detect

    func testNoChange() {
        XCTAssertNil(corrected(["jon"], "tell jon", "tell jon"))
    }

    func testEditToWordSpeechXDidNotInject() {
        // "work" wasn't injected → editing it is the user's own text, not a correction.
        XCTAssertNil(corrected(["jon"], "email jon at work", "email jon at home"))
    }

    func testIgnoresCommonHomophone() {
        XCTAssertNil(corrected(["there"], "go there now", "go their now"))
    }

    func testIgnoresTwoWordChange() {
        XCTAssertNil(corrected(["jon", "smith"], "call jon smith", "call john smyth"))
    }

    func testIgnoresUnrelatedReplacement() {
        XCTAssertNil(corrected(["cat"], "the cat", "the dog"))
    }

    func testIgnoresShortWords() {
        XCTAssertNil(corrected(["it"], "go to it", "go to is"))
    }

    // MARK: - Levenshtein

    func testLevenshtein() {
        XCTAssertEqual(TypeOverDetector.levenshtein("jon", "john"), 1)
        XCTAssertEqual(TypeOverDetector.levenshtein("kitten", "sitting"), 3)
        XCTAssertEqual(TypeOverDetector.levenshtein("same", "same"), 0)
    }
}

File: windows/SpeechX.Tests/TypeOverDetectorTests.cs
using SpeechX.Core;
using Xunit;

namespace SpeechX.Tests;

/// <summary>
/// Mirrors the macOS <c>TypeOverDetectorTests</c> so the two ports stay behaviourally identical.
/// </summary>
public class TypeOverDetectorTests
{
    private static string? Corrected(string[] injected, string baseline, string current)
        => TypeOverDetector.Correction(
            injected.Select(w => w.ToLowerInvariant()).ToHashSet(),
            baseline, current)?.Corrected;

    // MARK: - Should detect a correction

    [Fact]
    public void BasicSpellingFix()
        => Assert.Equal("john", Corrected(new[] { "tell", "jon", "about", "it" }, "tell jon about it", "tell john about it"));

    [Fact]
    public void PreservesCorrectedWordCasing()
        => Assert.Equal("John", Corrected(new[] { "jon" }, "hi Jon", "hi John"));

    // Regression: count-based diff must survive repeated words in the field.
    [Fact]
    public void CorrectedWordAlreadyPresentElsewhere()
        => Assert.Equal("ayush", Corrected(new[] { "ayewsh" }, "hi ayush ayewsh", "hi ayush ayush"));

    [Fact]
    public void InjectedWordRepeatsEditingOneOccurrence()
        => Assert.Equal("ayush", Corrected(new[] { "ayewsh" }, "ayewsh ayewsh", "ayush ayewsh"));

    // MARK: - Should NOT detect

    [Fact]
    public void NoChange()
        => Assert.Null(Corrected(new[] { "jon" }, "tell jon", "tell jon"));

    [Fact]
    public void EditToWordSpeechXDidNotInject()
        // "work" wasn't injected → editing it is the user's own text, not a correction.
        => Assert.Null(Corrected(new[] { "jon" }, "email jon at work", "email jon at home"));

    [Fact]
    public void IgnoresCommonHomophone()
        => Assert.Null(Corrected(new[] { "there" }, "go there now", "go their now"));

    [Fact]
    public void IgnoresTwoWordChange()
        => Assert.Null(Corrected(new[] { "jon", "smith" }, "call jon smith", "call john smyth"));

    [Fact]
    public void IgnoresUnrelatedReplacement()
        => Assert.Null(Corrected(new[] { "cat" }, "the cat", "the dog"));

    [Fact]
    public void IgnoresShortWords()
        => Assert.Null(Corrected(new[] { "it" }, "go to it", "go to is"));

    // MARK: - Levenshtein

    [Fact]
    public void Levenshtein()
    {
        Assert.Equal(1, TypeOverDetector.Levenshtein("jon", "john"));
        Assert.Equal(3, TypeOverDetector.Levenshtein("kitten", "sitting"));
        Assert.Equal(0, TypeOverDetector.Levenshtein("same", "same"));
    }
}
